### ISSUE TITLE
Authentication Headers not properly sent

### ISSUE THREAD

**Issue by wiwimaster** [10up/ElasticPress#1892]
Hi

SEE FOLLOW UP COMMENT FOR ADDITIONAL INFORMATION

I have successfully connected the ElasticPress Plugin (Version ) to an elastic.co ElasticSearch instance on AWS (tried multiple combinations of AWS, Google Cloud, or Azure, with versions 7.9 down to 6.8). I have various signals that the ElasticSearch Server is set up properly (see details below), but I keep getting a connection error, sometimes with strange reasons: 

`There is a problem with connecting to your Elasticsearch host. ElasticPress can try your host again, or you may need to change your settings. Response error: A valid URL was not provided.
`

The response error varied, including 400 and - this one is really interesting - a 200 OK "error":

![image](https://user-images.githubusercontent.com/19635773/93272879-d18cde00-f7b6-11ea-8099-d93e9a6c36ca.png)

I hope to get some help to get this resolved, and have included more information below.

## Issue:
with the error message in place, I cannot make use of the plugin, everything is greyed out:
![image](https://user-images.githubusercontent.com/19635773/93260729-11959600-f7a2-11ea-8c2f-5799a5e37192.png)

### Indexing
For some reason, I was able to send all pages to the index, which I can see both in the plugin and the elastic.co instance. I managed to do this by 1) including URL and auth info into the config.php as shown above and then 2) uninstall and reinstall the plugin, and immediately trigger an index:
![image](https://user-images.githubusercontent.com/19635773/93273148-78717a00-f7b7-11ea-9899-4d5add784dbb.png)

## Debug information

I used the Wordpress Debug Bar for ElasticPress to validate the connection, and I see three successful queries: 

![image](https://user-images.githubusercontent.com/19635773/93260506-caa7a080-f7a1-11ea-9a08-a41df9620e00.png)

it even adds config recommendations to another test I did with QBOX (and the error message remains the same here as well):

![image](https://user-images.githubusercontent.com/19635773/93266960-35110e80-f7ab-11ea-9003-4509d9d5e03e.png)

I've tested this on two different Wordpress 5.5.1 instances, both with elastic.co and QBOX, with no change in results.

In both cases, I have followed the best practice from other similar issues here and included

```
define( 'EP_HOST', 'https://[REMOVED].qb0x.com:31120/' );
define( 'ES_SHIELD', '[real_username]:[real_password]' );
add_filter( 'ep_format_request_headers', function ( $headers ) {
    $headers['Accept-Encoding'] = 'gzip';
    return $headers;
});
```

### CURL
I've also tried to CURL the elastic.co server from the webserver that runs the Wordpress instance with the plugin, which was successful:

```
:~$ curl -u elastic:PASSWORD https://HOST.cloud.es.io:9243
{
  "name" : "instance-0000000001",
  "cluster_name" : "17abc126f71445249c0ffb580885cf4f",
  "cluster_uuid" : "IcO3XnFTRhi76Xbr8xDFpQ",
  "version" : {
    "number" : "6.8.12",
    "build_flavor" : "default",
    "build_type" : "tar",
    "build_hash" : "7a15d2a",
    "build_date" : "2020-08-12T07:27:20.804867Z",
    "build_snapshot" : false,
    "lucene_version" : "7.7.3",
    "minimum_wire_compatibility_version" : "5.6.0",
    "minimum_index_compatibility_version" : "5.0.0"
  },
  "tagline" : "You Know, for Search"
}
```

Can you help me how to solve this?
thanks
Jan

**Comment by wiwimaster** [comment]
I dug a bit deeper and identified the root cause. Basically, the root cause can be traced down to the function get_elasticsearch_info() in Elasticsearch.php, which sends a request without authentication information, contrary to the other functions. See more details below:

The following code is leading to a false response. 

```
if ( ! empty( $es_info ) ) {
// Set ES info from cache.
$this->elasticsearch_version = $es_info['version'];
$this->elasticsearch_plugins = $es_info['plugins'];
}
```

In consequence, the following "else" statement requests the version & plugins but does not send authentication with it, in contrary to other queries: 

```
Headers: 
array(1) {
  ["Content-Type"]=>
  string(16) "application/json"
}

```

the response is obviously as follows:

```
Query Response Code: HTTP 401
Query Result: 
{
    "error": {
        "root_cause": [
            {
                "type": "security_exception",
                "reason": "action [cluster:monitor/nodes/info] requires authentication",
                "header": {
                    "WWW-Authenticate": [
                        "Bearer realm="security"",
                        "ApiKey",
                        "Basic realm="security" charset="UTF-8""
                    ]
                }
            }
        ],
        "type": "security_exception",
        "reason": "action [cluster:monitor/nodes/info] requires authentication",
        "header": {
            "WWW-Authenticate": [
                "Bearer realm="security"",
                "ApiKey",
                "Basic realm="security" charset="UTF-8""
            ]
        }
    },
    "status": 401
}
```


For comparison, other headers (of working queries) look like this (example below from /_cat/indices?format=json):

```
array(2) {
  ["Content-Type"]=>
  string(16) "application/json"
  ["Authorization"]=>
  string(82) "Basic ZGUuYmFja2VubWFjaHRnbHVlY2tsaWNoOmRoV1ZtSWJIY1dIWlBzSWNlV2VxR0ZnZ1dOVFlLaQ=="
}

```


As stated above, I have set ES_SHIELD in my Wordpress code properly. When I replace the above-mentioned IF statement with a if (true) and hard-coded values for version and plugins, everything works without any issues.

I'm not sure if this is a bug or just an advanced security setting, but I don't know how to overcome this and would be grateful if someone takes a look at this.

Best
Jan

**Comment by marcelmastel** [comment]
How did you solve?

**Comment by wiwimaster** [comment]
Hi Marcel

thanks for asking - I didn't. I was hoping to get some help here. As I didn't want to wait months for a reply and wanted to get going, I am now writing my own plugin to connect Wordpress to elasticsearch, which turns out is quite easy (and the official PHP client for Elasticsearch is waaaay better documented).

Good luck
Jan

**Comment by KalmanG** [comment]
@wiwimaster Is your solution working well? Any chance you would share your code? Having troubling connecting.

**Comment by wiwimaster** [comment]
@KalmanG, yes the solution works very well. You can find the API at https://www.elastic.co/guide/en/elasticsearch/client/php-api/current/index.html with all necessary descriptions to query, transform, update etc elasticsearch. 

A simple query would look like this (more or less copy&paste from the "quick start" example in the mentioned doc above):

```
//load the API and connect to elasticsearch 
require 'vendor/autoload.php';
use Elasticsearch\ClientBuilder;
$client = ClientBuilder::create()->build();

//define query parameters, following elasticsearch official query DSL
$params = [
    'index' => 'my_index',
    'id'    => 'my_id',
    'body'  => ['testField' => 'abc']
];

//execute the query
$response = $client->index($params);

//print results
print_r($response);
```

In terms of effort, it totally depends on what you want to do. I have used it for writing a recommendation engine based on existing posts, but made it a bit more specific & feature-rich vs. the elasticPress version. I have spent a lot of time on it, but it always was a steady progress, with little time wasted.

**Comment by brandwaffle** [comment]
@KalmanG @wiwimaster not sure if perhaps you have custom filters applied or another plugin is somehow conflicting, but with a base install this is the query that is issued for /nodes/plugins (note the authorization string is passed).

![image](https://user-images.githubusercontent.com/663228/104344671-4ac0e700-54b2-11eb-9cc4-05ade530a1d8.png)

http://10up.github.io/ElasticPress/includes_classes_Elasticsearch.php.html#line542 shows that the basic auth header is added for all queries, and again, when testing with a base install we get the above result.

**Comment by wiwimaster** [comment]
Hi @brandwaffle great to get some feedback at last after 4 months. 

If you would have read my text above thoroughly, you'll see that at the time of opening this issue, there was no auth sent, so if you send it now, it is great but too late in my case. Also, I don't know many people who intend to use a plugin on a "clean" install only, so I expect plugins to also work with other plugins. Of course, I am aware of plugin incompatibilities, that's why I tried it on two completely different (up-to-date) WordPress installations - with the same result.

In my case - this is history. After not getting any help at all and investing hours in finding and fixing this issue, I decided to move to a custom-built solution, which works significantly better for me. I'm aware this is not the best solution for everyone, and I'm glad to hear that this issue seems to be fixed now.
